proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -983,6 +983,743 @@ lemma replyRemoveTCB_corres:
+                                                   and (\<lambda>s. (replies_of' s |> replyNext) rp = None)
+                                                   and (\<lambda>s. \<forall>p'. replyPrevs_of s p' \<noteq> Some rp)
+                                                   and (\<lambda>s. \<forall>p'. scReplies_of s p' \<noteq> Some rp)"
+                                       in corres_inst)
+                                apply (rule corres_symb_exec_r_sr_strong) (* replyPrev at rp = None *)
+                                   apply (rule corres_guard_imp)
+                                     apply (rule replyUnlinkTcb_corres)
+                                    apply (clarsimp simp: valid_objs_valid_tcbs elim!: pred_tcb_weakenE)
+                                   apply simp
+                                  apply (simp add: cleanReply_def)
+                                  apply (rule_tac Q'="\<lambda>_ s. reply_at' rp s \<and> (replies_of' s |> replyNext) rp = None
+                                                            \<and> (\<forall>p'. replyPrevs_of s p' \<noteq> Some rp)
+                                                            \<and> (\<forall>p'. scReplies_of s p' \<noteq> Some rp)"
+                                         in sr_inv_ul_bind[rotated])
+                                    apply (rule updateReply_sr_inv)
+                                     apply (clarsimp simp: reply_relation_def)
+                                    apply (intro conjI impI allI)
+                                    apply (erule sc_replies_relation_replyNext_None; clarsimp)
+                                    apply (clarsimp simp: obj_at'_def projectKOs opt_map_red)
+                                   apply clarsimp
+                                   apply (wpsimp wp: updateReply_wp_all)
+                                   apply (clarsimp simp: obj_at'_def projectKOs objBits_simps ps_clear_upd opt_map_red)
+                                   apply (rename_tac s s' reply' sc')
+                                   apply (intro conjI allI; clarsimp split: if_split_asm simp: projectKOs)
+                                   apply (rename_tac scp' sc'')
+                                   apply (drule_tac x=scp' in spec[where P="\<lambda>x. scReplies_of _ x \<noteq> Some rp"])
+                                   apply (clarsimp simp: opt_map_red)
+                                  apply (clarsimp simp: sr_inv_def updateReply_def)
+                                  apply (clarsimp simp: setReply_def getReply_def getObject_def
+                                                        setObject_def split_def objBits_simps'
+                                                        updateObject_default_def in_monad fail_def
+                                                        in_magnitude_check obj_at_simps return_def
+                                                        loadObject_default_def ARM_H.fromPPtr_def
```

Comment:
It’s not much work to put this lemma in the arch locale, but I am also concerned about going in and out of locale like that. And if we want to optimise locale transition, say, at the theory level, it becomes more work that it might not be worth spending time on right now, and we might as well wait for doing arch-split properly at some point.